### DISCUSSION TITLE
光驱偏移值对抓轨的实际影响

### DISCUSSION THREAD

**Discussion by giscus** [ZexWoo/Giscus#29] [Announcements]
# 光驱偏移值对抓轨的实际影响

如果你是一个有一定经验的 CD 抓轨玩家，你一定会注意到「偏移值」这个东西。那么，它是如何影响我们的抓轨结果的呢？本文将为你揭开这个秘密。

https://zexwoo.blog/posts/research/offsets-affect-ripping/

**Comment by Xin-github404** [comment]
请问，有哪些光驱支持逾读呢？

**Reply by ZexWoo** [reply]
文中使用的 BENQ CDRW 52X32，以及 5232X，都是支持的，其他的，不太清楚。

**Reply by Xin-github404** [reply]
好的，谢谢博主。你说的那两款光驱市面上有点难找😂。
我手里有三台光驱，它们貌似都不支持逾读，但是我用它们抓轨的时候发现一个问题。
抓某些碟片的时候，即使开了通读选项，log里面也不报错。抓其他碟片时，开启通读选项，log里面就会报错 (原因是出现可疑位置，且首音轨不能验证，末音轨有错误)，而且我观察了一下，抓轨过程，抓末音轨显示「同步错误」。
我就疑问，如果光驱不支持逾读，开启通读选项，按理说log文件都会报错。所以，上述的情况是什么原因呢？

**Reply by ZexWoo** [reply]
可能与 CD 本身的数据情况有关联。如果不支持，那就不开，反正不影响抓轨和聆听。

**Comment by Lohym** [comment]
> CD：doriko - 花束 ～the best of doriko feat. 初音ミク～；

~~**doriko好评！**~~
这里有一个[光驱DAE特性的数据库](https://www.daefeatures.co.uk/search)，可以查询光驱的读写偏移、是否支持踰读、隐藏音轨抓取等等。不过由于数据库中有详细特性的大部分貌似都是比较老型号的，加上都是用户上传的数据，所以也就**只能当个参考，不完全保真**。如果对自己手里的光驱做了详细测试，也可以注册并向数据库上传。
关于踰读，我也拿自己手上的光驱做了一个小测试：
**Dell为便携式USB接口光驱，EW164B和52X32自带光驱盒，其他均为5.25寸内置光驱且均使用易驱线进行测试，驱动器是否缓存音频按照EAC检测结果设置。**
**测试所用CD：** `HOYO-MiX - CD 03 Battles of Liyue 璃月鏖战录`，使用这张是因为它是已知在+6偏移量且不支持踰读光驱在错误开启踰读后会出现同步错误的CD（这张CD仅不到12分钟）。
**测试结果如下：**
| 属性\型号 | BenQ 652M | BenQ CW5232 | Dell DW316 | HL-DT-ST GH30N | Lite-on iHDS118 7 | Pioneer DVR-220CHV | Sony DDU1681S-BK | BenQ EW164B |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| 类型 | CD-ROM | CD-RW | DVD+-RW | DVD+-RW | DVD-ROM | DVD+-RW | DVD-ROM | DVD+-RW |
| 读取偏移 | +6  | +6  | +6  | +102  | +6  | +6  | +6  | +618  |
| 踰读(EAC) | 导入 | 导入 | 无 | 导入和导出 | 无 | 无 | 导入 | 导入 |
| 导入(NDS) | 无 | 142 | 无 | 144 | 1 | 无 | 140 | 142 |
| 导出(NDS) | 无 | 6749 | 无 | 无 | 无 | 无 | 无 | 无 |
| 开踰读CRC | CF2506BE | CF2506BE | CF2506BE | 561F73BF | CF2506BE | CF2506BE | CF2506BE | F232D5F4 |
| 关踰读CRC | 561F73BF | 561F73BF | 561F73BF | 561F73BF | 561F73BF | 561F73BF | 561F73BF | 561F73BF |
| 开C2和踰读CRC | 561F73BF | 561F73BF | 561F73BF | 未抓取 | 561F73BF | 561F73BF | 561F73BF | 未抓取 |

**[所有Log](https://pan.baidu.com/s/1GUIL69_t43t-N_YiUsQD7g?pwd=ODDT)**
**踰读(EAC)**：EAC对踰读的检测结果。
**导入(NDS)与导出(NDS)**：`Nero DiscSpeed`的高级DAE特性对光驱读取导入和导出区的检测结果，单位为Sector即扇区（也就是博客中说的块），每个扇区有588个Sample，这个Sample也就是偏移值的单位。
**CRC**：抓取Log中的整轨CRC，`561F73BF`为正常抓取无错误，`CF2506BE`和`F232D5F4`的CRC为在抓取末尾（进度99.9%）出现同步错误并在log中存在可疑位置（在最后一首歌的最后一秒钟）的情况。“未抓取”是因为`HL-DT-ST GH30N`在EAC和`Nero DiscSpeed`中都没有检测到C2功能（EAC报告“无法找到匹配的C2读取模式”，`Nero DiscSpeed`里与C2相关的功能直接被禁用了，神奇的是使用`dbpoweramp`依然能在正确的位置检测到C2），`BenQ EW164B`没有抓是单纯因为~~懒~~觉得没有必要了。

<details>
<summary>“561F73BF”（正常）与“CF2506BE”（同步错误）文件的区别</summary>

上面为“CF2506BE”，丢失了部分采样，下面为“561F73BF”。
（嗯，丢的可比6个多多了。~~怎么和goldwave还有不一样的，EAC的检测是不是不太准确~~）
    
![对比图](https://i.postimg.cc/KjdGM9sr/Snipaste-2025-04-18-00-59-29.png)

</details>

##### 那么根据以上信息可看出：
1. 6个光驱偏移量为+6且EAC检测均不支持读取导出区，打开踰读后均出现同步错误，且文件CRC相同。偏移量为+102的光驱EAC检查支持读取导出区且打开踰读与关闭时得到的文件CRC相同，也和其他光驱关闭踰读抓取的文件CRC相同。偏移量为+618的抓取结果在log里的cuetools那里可见，只和偏移量为+6的相差一个sample。
2. 偏移量为+6的光驱在同时开启C2和踰读后，抓取正常结束没有出现同步错误，文件与不开踰读时的CRC相同。
3. 因为错误设置踰读进行抓取而产生的错误文件会在末尾填充更多的、远大于该光驱偏移量的静音。
4. EAC与NDS的踰读检测结果不完全一致，EAC的检测结果与开关踰读的CRC一致。
##### 根据以上信息可以进行推断：
1. 如果一款光驱在EAC中不支持踰读，那么该光驱在打开踰读关闭C2的情况下抓取会出现同步错误CD所产生的文件，应该与和该光驱偏移量相同的、在EAC中同样设置下抓取同样CD所得到的错误文件完全一致。（需进一步验证）
2. 根据博客中的结论：`如果光驱本身没有「踰读（overread）」能力，EAC 中也没有开启此功能，则正偏移值越大，就有损失越多尾部采样的风险`，因此CD节目区的最后102到最后6个Sample应该是静音（否则CRC应不同）。
3. 鉴于在正常读取情况下可读取数据的末尾已经是静音，后续数据为静音的可能性也较大，因此符合`HL-DT-ST GH30N`抓取的文件均正确且CRC相同这一现象，`HL-DT-ST GH30N`拥有读取导出区的能力。（买了5232WI，应该就是52X32，开踰读抓出来也是`561F73BF`，也就是说最后102个sample都是静音了）
4. 若使用偏移量比+6更大的不支持踰读的光驱在开启踰读关闭C2的情况下抓取这张CD，丢失的采样会比偏移量为+6的光驱更多（也就是说文件会在末尾填充更多的静音）。（事实与我想象的有一定差距，偏移量+618和+6居然只相差一个sample）
5. 根据[Wiki](https://wiki.hydrogenaud.io/index.php?title=EAC_Drive_Options#Drive_is_capable_of_retrieving_C2_error_information)和[dbpoweramp作者](https://www.dbpoweramp.com/spoons-audio-guide-cd-ripping.htm)的说法，C2开启后EAC不会再进行“安全重读”（至少读2次）而仅依靠C2是否报错（也就是说会变成读一次+检查C2是否报错），由于我使用的CD几乎没有什么划痕，因此可以认为应该不会有C2错误（事实上在这次测试中，在“开踰读+C2”的设置下也没有看到报错灯亮过），也就是说在读一次的情况下不会出现同步错误。（不过这里也不能排除C2本身影响抓取的可能）

这个测试不太完整，主要还是因为手里的光驱属性差异有点小了，部分推断还需要使用其他的光驱进行验证。不过这里又有一些新的问题：
1. 同步错误是什么？（在[EAC官网](https://www.exactaudiocopy.de/en/index.php/overview/basic-technology/extraction-technology/#:~:text=Only%20if%20a%20read%20error,blocks%20of%2023%20total%20blocks.)有解答）
2. 为什么在有重读（即关C2）的时候会出现同步错误，而实际只读一次（即开C2）的时候文件又没有错误？
3. 为什么不支持踰读的光驱开启踰读后会在末尾处提早（提早是指相比不开踰读能读取到的位置）出现同步错误？
4. 究竟应该如何正确地测试踰读功能？最简单省力的方法是**使用多张CD进行重复的检测，因为不同CD的检测结果可能发生变化**（**需要注意的是，偏移值即EAC中所谓“偏移校正值”为正的需要读取导出区，偏移值为负的则需要读取导入区**。而令人感到无语的是，目前偏移值数据库中为正的占大部分，但其中部分光驱仅有读取导入区的能力，也就是说不能使用踰读功能）。而如果要做实际有无读取能力的测试的话，说来也“简单”，检测光驱能否正确读取导入导出区内的数据即可。导入区的话以前[有人提到过](https://hydrogenaud.io/index.php/topic,38551.msg340297.html#msg340297)某些CD的第一轨有部分在导入区，有办法进行抓取。导出区我推测可以自己超刻音乐CD来进行测试，这种需要可以超刻的光驱和CD。都需要特定的东西，说实话都挺麻烦的……

<details>
<summary>不同CD通读检测结果不同</summary>

**甚至上面两张还都是EAC offset测试CD，只是写入偏移不同而已**

![通读测试](https://i.postimg.cc/fW6fJ62W/1.png)
</details>

要解答有的疑问感觉有点困难。EAC的踰读检测结果和`Nero DiscSpeed`不一致，检测方法也不太相同（EAC的踰读检测可以使用多张CD重复检测，而`Nero DiscSpeed`需要使用该软件先进行刻录制作专用的检测CD才能进行检测，其他CD则无法使用该功能），而到底谁是正确的也难以分辨。以这次测试的结果来说，EAC实际上抓取时的结果印证了EAC对导出区读取能力的检测结果是正确的。（写到最后我发现前面应该是我有误解了，EAC的踰读功能可能与能否读取导入或导出区根本没有关系，因为丢失的sample应该没有在CD的物理导入或导出区内。）

**为啥对这东西这么在意**？ 之前错误的开启了踰读进行抓取，碰巧遇到了同步错误，甚至让我误以为是CD品质的问题，真是令我印象深刻。说实话为了抓轨都用上EAC了，既然它又有这个功能，我觉得就应该尽可能完整的把数据给保存下来（~~说实话这可能是强迫症犯了~~）。另外就是最近对这些不同光驱的功能特性比较感兴趣。

**为什么测试会有用到C2**？ 我最开始是以为C2是有或没有的功能，也以为EAC的检测是准确的，结果没想到这东西也是个[有好有差的功能](https://www.dbpoweramp.com/spoons-audio-guide-cd-ripping.htm)，而EAC的检测也并不准确……（算是误打误撞，不过在不支持踰读的光驱上“开踰读+C2”已经算是错上加错了）**鉴于EAC用了这个后就不重读了，所以不管有无都不建议开C2**。而且EAC和dbpoweramp都不能准确地评估C2（不过dbpoweramp开启C2也还是会使用重读，这点算是比EAC好了），**它们只能判断光驱有没有C2功能，而不会判断它是否准确**（如果真想测试光驱C2能力可以试试[这个](https://www.exactaudiocopy.de/en/index.php/other-projects/dae-quality/)）。

另外，[一部分的隐藏第一音轨（Hidden Track One Audio，HTOA）应该与踰读功能是没有关系的](https://hydrogenaud.io/index.php/topic,38551.msg340297.html)，对于这个功能我也用[这个方法](https://www.daefeatures.co.uk/htoa)做了测试，实际上看到了四种不同的情况……

PS：溯因那里引用的第一句话`校验码算法会忽略第一个音轨开头的 2939 个采样（不到 5 块，略大于 0.013 秒的音频）`，这里的0.013秒是说的单个块的长度吧。
PS2：买了52X32后发现，这货抓取速度是真的快得离谱……在同样设置的情况下速度可以甩开我手上的其他光驱几条街了

**Reply by ZexWoo** [reply]
辛苦了！做了这么多测试。我看了一下引文，英文原文确实是指的一个块的时长。

很巧，我也有这套专辑，就拿手上的光驱抓了一下，结果得到了比较有意思的结果。

[度盘链接](https://pan.baidu.com/s/1A0yIqdbzKmXgiA3y1jlTQg?pwd=6d71)

简单来说，我测试了整张抓取和是否开踰读的四次测试，结果每次第三轨的 CRC 都是 `786EDC50`，和你抓出来的不太一样。音频文件我也上传了，你可以下载以后比对一下看看哪里不一样。

**Reply by Lohym** [reply]
额，52X32是支持踰读的，所以它不管怎么抓这张结果都是正常的。
如果您想抓出有问题的结果的话应该需要一个偏移值为正的且EAC检测不支持读取lead-out的光驱。

我用EAC的`按CUE目录文件分割WAV文件`-`单个间隙`分割CRC为`561F73BF`的整轨后，用EAC的`处理WAV文件`打开分割好的第三轨，CRC就是`786EDC50`，也就是说这其实就是正常的抓取结果。所以对比结果其实就和我上面的那个截图是一样的。

我各上传了一个偏移量+6和+618的踰读失败了的整轨音频。[链接](https://pan.baidu.com/s/1MJRj8Xy1OasFsxRrFlBLpA?pwd=ODDT)

**Reply by Xin-github404** [reply]
感谢层主的调研！所以说，从能完美复制CD数据并泛用至所有CD的角度来说，该如何检测光驱的各项属性 (逾读、读隐藏轨等) 呢？能不能找到或制造一张CD，能通用地检测所有光驱呢？有人觉得C2纠错的优先级是高于重复读取的，因为正常光驱读一次是不会出错的。我觉得网上讨论这些的情况还是比较少，或许制造光驱的工程师更懂些。

**Reply by Lohym** [reply]
要想用一张CD来检测有点太难了，毕竟有些功能我们不知道EAC的检测根据是什么。

踰读检测既然观测到了在同一个光驱上用不同CD进行检测得到的结果会发生变化，最好用多张正版CD检测，得到的结果才比较有说服力。当然也可以像我上面一样，拿一张已知错误开启踰读抓取时会出现错误的CD来进行抓取检测（这个大概是可以通用的，至少在具有相同偏移量的光驱之间能够通用的可能性比较高），不如说这才是在实际上测试了是否能正常进行踰读。偏移量为负的光驱的情况我不是很清楚，不过这种光驱挺少见的，毕竟偏移量数据库里偏移量为负的光驱仅占总数的不到四十分之一Orz……

第一隐藏音轨由于我没有具有这种特性的正版CD，就只能根据上文所说的方法自己刻录CD-R进行测试了。不过我觉得最好还是用正版CD进行测试，因为我不知道它们之间是否还有什么其他不同的地方。上述机型支持抓取第一隐藏音轨的有`HL-DT-ST GH30N`、`Lite-on iHDS118 7`、`Sony DDU1681S-BK`（这三个还都是sata接口的），其他的要么抓出来在中间有缺陷，要么抓出来第一隐藏音轨是空的，要么碟放进去后直接把EAC卡死。
另外，普通的第一隐藏音轨在抓整轨的cue文件里可以看出来和其他CD的不同，这种CD在第一音轨具有`INDEX00`，其他CD虽然都会在第一音轨有2秒间隙但是在cue文件里不会出现（这点我也是刚刚看到博主的分轨log里有在第一音轨里显示pregap2秒才想起来的）。

C2这个东西我理解得不是很透彻，我不太清楚这个功能是只是会进行校检是否有错误，还是能同时通过它进行修复，不同软件的实现可能也不太一样（为什么这么说，因为作为CD刻录质量检测机而知名的BenQ DW1640——也就是BenQ EW164B内部的光驱型号，使用EAC和dbpoweramp都无法检测C2，EAC提示“无法找到匹配的 C2 读取模式”，dbpoweramp的检测条直接不动，这显然不太对）。但我**不推荐在EAC里使用C2**，原因上面已经说过了。

<details>
<summary>我用自己刻的C2测试碟在52X32上进行测试时依然能检测到错误并进行修复</summary>

注意设置关闭了`驱动器可以缓冲音频数据`（实际上这个光驱是会缓冲音频的）并且开启了C2，但是数据并不是完全对得上的，不知道是刻录软件的问题还是什么（用的Nero Burning ROM），在前3轨中总会在最前面的位置有6个sample的重复或者缺失，由于是用马克笔直接在CD上涂抹的，第4轨以后错误太多以至于下面的Suspicious position的地方几乎都是没有声音的。
    
    Exact Audio Copy V1.8 from 15. July 2024
    
    EAC extraction logfile from 24. April 2025, 2:07
    
    DeZI:R / Test-Wa
    
    Used drive  : ATAPI   CDRW 52X32   Adapter: 1  ID: 1
    
    Read mode               : Secure
    Utilize accurate stream : Yes
    Defeat audio cache      : No
    Make use of C2 pointers : Yes
    
    Read offset correction                      : 738
    Overread into Lead-In and Lead-Out          : Yes
    Fill up missing offset samples with silence : Yes
    Delete leading and trailing silent blocks   : No
    Null samples used in CRC calculations       : Yes
    Used interface                              : Native Win32 interface for Win NT & 2000
    Gap handling                                : Not detected, thus appended to previous track
    
    Used output format : Internal WAV Routines
    Sample format      : 44.100 Hz; 16 Bit; Stereo
    
    
    TOC of the extracted CD
    
         Track |   Start  |  Length  | Start sector | End sector 
        ---------------------------------------------------------
            1  |  0:00.00 |  2:07.10 |         0    |     9534   
            2  |  2:07.10 |  3:54.06 |      9535    |    27090   
            3  |  6:01.16 |  3:48.12 |     27091    |    44202   
            4  |  9:49.28 |  4:12.18 |     44203    |    63120   
            5  | 14:01.46 |  3:46.39 |     63121    |    80109   
            6  | 17:48.10 |  5:46.09 |     80110    |   106068   
            7  | 23:34.19 |  5:01.03 |    106069    |   128646   
            8  | 28:35.22 |  4:41.36 |    128647    |   149757   
            9  | 33:16.58 |  3:52.72 |    149758    |   167229   
           10  | 37:09.55 |  3:39.42 |    167230    |   183696   
           11  | 40:49.22 |  4:04.00 |    183697    |   201996   
           12  | 44:53.22 |  3:53.21 |    201997    |   219492   
           13  | 48:46.43 |  4:57.34 |    219493    |   241801   
           14  | 53:44.02 |  6:54.14 |    241802    |   272865   
           15  | 60:38.16 |  4:12.36 |    272866    |   291801   
           16  | 64:50.52 |  3:39.24 |    291802    |   308250   
    
    
    Track  1
    
         Filename E:\ODD-Test\Benq_52X32\NoCache-C2\01.wav
    
         Peak level 99.8 %
         Extraction speed 18.3 X
         Track quality 99.8 %
         Copy CRC CB87096B
         Copy OK
    
    Track  2
    
         Filename E:\ODD-Test\Benq_52X32\NoCache-C2\02.wav
    
         Peak level 99.9 %
         Extraction speed 20.9 X
         Track quality 100.0 %
         Copy CRC 3EDEA168
         Copy OK
    
    Track  3
    
         Filename E:\ODD-Test\Benq_52X32\NoCache-C2\03.wav
    
         Peak level 100.0 %
         Extraction speed 22.6 X
         Track quality 99.9 %
         Copy CRC 3A0E5B26
         Copy OK
    
    Track  4
    
         Filename E:\ODD-Test\Benq_52X32\NoCache-C2\04.wav
    
         Suspicious position 0:00:27
         Suspicious position 0:00:44
         Suspicious position 0:01:01
         Suspicious position 0:01:08
         Suspicious position 0:01:13
         Suspicious position 0:01:18
         Suspicious position 0:01:27
         Suspicious position 0:01:37
         Suspicious position 0:01:44
         Suspicious position 0:02:03
         Suspicious position 0:02:18
         Suspicious position 0:02:33
         Suspicious position 0:02:46
         Suspicious position 0:02:49
         Suspicious position 0:02:52
         Suspicious position 0:02:54 - 0:02:57
         Suspicious position 0:03:00 - 0:03:01
         Suspicious position 0:03:05
         Suspicious position 0:03:07
         Suspicious position 0:03:18 - 0:03:19
         Suspicious position 0:03:27 - 0:03:28
         Suspicious position 0:03:36
         Suspicious position 0:03:39
         Suspicious position 0:03:47
    
         Peak level 99.4 %
         Extraction speed 3.0 X
         Track quality 95.8 %
         Copy CRC AB468B51
         Copy finished
    
    Track  5
    
         Filename E:\ODD-Test\Benq_52X32\NoCache-C2\05.wav
    
         Copy aborted
    
    Track  6
    
         Copy aborted
    
    Track  7
    
         Copy aborted
    
    Track  8
    
         Copy aborted
    
    Track  9
    
         Copy aborted
    
    Track 10
    
         Copy aborted
    
    Track 11
    
         Copy aborted
    
    Track 12
    
         Copy aborted
    
    Track 13
    
         Copy aborted
    
    Track 14
    
         Copy aborted
    
    Track 15
    
         Copy aborted
    
    Track 16
    
         Copy aborted
    
    There were errors
    
    End of status report
    
    ---- AcoustID Plugin V1.2.0
    
    Submitting 3 results to AcoustID
    
    AcoustID successfully submitted
    
    
    ==== Log checksum 0742A3F1549139A27AE7EFA2E4984EE298391465971556ABF823E539878DA91B ====
    

</details>

上面测试的这几个光驱之中`HL-DT-ST GH30N`算是比较理想的一个了，抓取速度中等偏上（是在关闭`驱动器可以缓冲音频数据`选项的情况下，因为我用4个不同软件检测都显示不缓冲音频）、支持踰读和抓取第一隐藏音轨、又是sata接口、价格应该也不是很贵（我是在线下店偶然拿到的二手），不过纠错能力我不是很清楚。

**Reply by ZexWoo** [reply]
原来你是整轨抓的，难怪 CRC 不同，我没细看。

**Reply by Lohym** [reply]
不好意思我自己抓的时候是整轨派（orz

而且整轨抓取的话速度可能相比分轨会快那么一丁点，从您正文中doriko的那个log里可以看到部分音轨的抓取速度是比前一个音轨的更低的，但是抓整轨的时候途中速度基本上是单调递增的，貌似整轨的纠错只会在抓取完成的最后进行（因为我正常情况下抓整轨从来没有看到途中纠错灯亮过。除非是出现同步错误，也就是用不支持抓取HTOA的光驱抓HTOA碟的时候），但分轨会在当前音轨抓取完后立刻进行。

<details>
<summary>不过如果貌似错误太多的话依然会在抓取时亮纠错灯</summary>
这也是为什么上面那个C2测试碟抓取时没有检测间隙，因为会读不出来让EAC卡死。
    
![图](https://i.postimg.cc/DwmW2hRT/Snipaste-2025-04-14-11-35-56.png)

</details>

**Reply by LzNOHentai** [reply]
大佬您好，请教一下目前抓取CD您推荐购入什么型号的光驱呢？目前我是ASUS 16D1HT包括BD DVD CD在内一把梭，还能抓UHD BD个人挺满意的，就是抓CD的速度确实不尽如人意。最理想的情况是能够在特性齐全的情况下速度越快越好，不过如果没有方便购入、保养状况良好的型号可选的话，至少抓取速度上希望有比较大的提升。
我用daefeatures检索，并把相关型号在咸鱼上检索，姑且有HL-DT-ST	CD-RW GCE-8526B这样的型号还有上古拆机货，但抓取速度就不知道怎么样了。大佬是否有什么推荐的型号？

**Reply by ZexWoo** [reply]
https://yasra.notion.site/yasra/f68f87af9b78476ea3695ed27808dc84 我的友人有搜集，看这个就好了。
